renderdoc/driver/shaders/dxbc/dxbc_debug.h

```diff
@@ -147,6 +147,31 @@ struct GlobalState
   std::map<SampleEvalCacheKey, ShaderVariable> sampleEvalCache;
 };
 
+class DebugAPIWrapper
+{
+public:
+  virtual void SetCurrentInstruction(uint32_t instruction) = 0;
+  virtual void AddDebugMessage(MessageCategory c, MessageSeverity sv, MessageSource src,
+                               std::string d) = 0;
+
+  virtual bool CalculateRcp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateRsq(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateExp(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateLog(const ShaderVariable &input, ShaderVariable &output) = 0;
+  virtual bool CalculateSinCos(const ShaderVariable &input, ShaderVariable &output1,
+                               ShaderVariable &output2) = 0;
```

Comment:
This seems like a good candidate to merge into a single `CalculateMath` function or something like that with the `DXBC::OpcodeType` as a parameter. In your D3D11 implementation it forwards onto a single function anyway with a different string for performing the operation itself, which could be replaced by a switch(). I would expect the same thing to happen for any other implementation so we might as well codify that in the interface and make it easier to extend if we decide we want to run other operations on the GPU.

(In the past I've run into driver/GPU specific behaviour around certain edge cases so I could see wanting to move other operations either entirely onto the GPU or doing little feature-detect type things to see how the GPU reacts to edge cases with denormals and NaN/Inf to adjust the emulation)

Reply:
Merged these into a single function.